racket-cas/racket-cas.rkt

```diff
@@ -1998,6 +2001,183 @@
   )
 
+(define (extract-related-operands u x)
+  (when debugging? (displayln (list 'extract-related-operands u x)))
+  (define (non-free-of-x? u)
+    (and (not (free-of u x)) (not (equal? u x))))
+  (define (non-free-of-x-or-empty-list u)
+    (if (non-free-of-x? u) `(,u) '()))
+  (math-match u
+              [r '()]
+              [y '()]
+              [(Sum _) '()]
+              [(Prod us) (filter non-free-of-x? us)]
+              [(app: _ us)
+               (match us
+                 [(list u) (non-free-of-x-or-empty-list u)]
+                 [(list u v)
+                  (cond
+                    [(and (non-free-of-x? u) (non-free-of-x? v)) `(,u ,v)]
+                    [(non-free-of-x? u) `(,u)]
+                    [(non-free-of-x? v) `(,v)]
+                    [else '()]
+                  )]
+                 [_ (error "more than two operands")]
+                 )]
+              [_ (error "no case matched.")]))
+
+(module+ test
+  (check-equal? (extract-related-operands (⊗ (Exp x) x) x) '((expt @e x)))
+  (check-equal? (extract-related-operands (⊗ 3 (Sqr y)) y) '((expt y 2)))
+  )
+
+(define (subst-candidates-lst u x)
+  (define (x-subst-candidates u)
+    (subst-candidates-lst u x))
+  (define (extract-x-related-operands u)
+    (extract-related-operands u x))
+  (define new-ops (extract-x-related-operands u))
+  (if (empty? new-ops)
+      '()
+      (append new-ops
+              (foldl append '()
+                     (map x-subst-candidates new-ops))
+              )))
+
+(define (subst-candidates u x)
+  (list->set (subst-candidates-lst u x)))
+
+(module+ test
+  (check-equal? (subst-candidates (⊗ (Exp x) x) x) (set '(expt @e x)))
+  (check-equal? (subst-candidates (⊗ (Exp (Sqr (Cos x))) x) x)
+                (set '(expt @e (expt (cos x) 2)) '(expt (cos x) 2) '(cos x)))
+  )
+
+(define (subst-with-symbol u x v [symbol 'symbol])
+  (define solution (roots (⊖ v symbol) x))
+  (if (empty? solution)
+      (subst u v symbol)
+      (subst u x (subst (first solution) @n 0))
+  ))
+
+(module+ test
+  (check-equal? (subst-with-symbol (⊗ (Exp x) x) x '(expt @e x)) '(* symbol (ln symbol)))
+  )
+
+(define (integrate-basic-failed? u)
+  (equal? u 'integrate-basic-failed))
+
+(define (try-subst-integrate u x v)
+  (when debugging? (displayln (list 'try-subst-integrate u x v)))
+  (define sym 'dummy-symbol)
+  (define substed (normalize (expand (subst-with-symbol (⊘ u (diff v x)) x v sym))))
+  (if (not (free-of substed x))
+    #f ; subst failed
+    (with-handlers ([integrate-basic-failed? (lambda (u) #f)]) ; integrate-basic failed
+    (subst (integrate-basic substed sym) sym v)
+      )
+  ))
+
+(module+ test
+  (check-equal? (try-subst-integrate (⊗ (Exp x) x) x '(expt @e x)) '(+ (* -1 (expt @e x)) (* (expt @e x) x)))
+  (check-equal? (try-subst-integrate (Sqr (⊕ x 1)) x '(+ x 1)) '(* 1/3 (expt (+ 1 x) 3)))
+  )
+
+(define (definite-integrate u x a b)
+  (define integral (integrate u x))
```

Comment:
; symbolic computation of the definite integral of u wrt to x with limits a and b.
Is the limits a and b meant to be finite - or is ≠inf.0 allowed?

Reply:
Only the basic cases are considered in the PR.